optlang/glpk_interface.py

```diff
@@ -585,45 +586,37 @@ def objective(self, value):
 
     @property
     def primal_values(self):
-        primal_values = collections.OrderedDict()
-        is_mip = self._glpk_is_mip()
-        for index, variable in enumerate(self.variables):
-            if is_mip:
-                value = glp_mip_col_val(self.problem, index + 1)
-            else:
-                value = glp_get_col_prim(self.problem, index + 1)
-            primal_values[variable.name] = variable._round_primal_to_bounds(value)
-        return primal_values
+        # round primals
+        primal_values = [variable._round_primal_to_bounds(primal)
+                         for variable, primal in zip(self.variables, self._get_primal_values())]
+        return collections.OrderedDict(
+            zip(self._get_variables_names(), primal_values)
+        )
 
-    @property
-    def reduced_costs(self):
+    def _get_primal_values(self):
+        if self._glpk_is_mip():
+            # no vector function (element wise)
+            return [glp_mip_col_val(self.problem, index + 1)
+                    for index in range(len(self.variables))]
+        else:
+            return get_col_primals(self.problem)
+
+    def _get_reduced_costs(self):
         if self.is_integer:
             raise ValueError("Dual values are not well-defined for integer problems")
-        reduced_costs = collections.OrderedDict(
-            (var.name, glp_get_col_dual(self.problem, index + 1)) for index, var in enumerate(self.variables)
-        )
-        return reduced_costs
+        return get_col_duals(self.problem)
 
-    @property
-    def constraint_values(self):
-        dual_values = collections.OrderedDict()
-        is_mip = self._glpk_is_mip()
-        for index, constraint in enumerate(self.constraints):
-            if is_mip:
-                value = glp_mip_row_val(self.problem, index + 1)
-            else:
-                value = glp_get_row_prim(self.problem, index + 1)
-            dual_values[constraint.name] = value
-        return dual_values
+    def _get_constraint_values(self):
+        if self._glpk_is_mip():
+            # no vector function (element wise)
+            return [glp_mip_row_val(self.problem, index + 1) for index in range(len(self.constraints))]
+        else:
+            return get_row_primals(self.problem)
```

Comment:
Same as before. Already handled by the swiglpk function. No need for glp_mip_row_val here as this is used already anyways.